app/services/archive/frozen_cohort_user.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module Archive
+  class FrozenCohortUser < ::BaseService
+    include Archive::SupportMethods
+
+    def call
+      check_user_can_be_archived!
+
+      data = Archive::UserSerializer.new(user).serializable_hash[:data]
+
+      ActiveRecord::Base.transaction do
+        relic = Archive::Relic.create!(object_type: user.class.name,
+                                       object_id: user.id,
+                                       display_name: user.full_name,
+                                       reason:,
+                                       data:)
+        destroy_user! unless keep_original
+        relic
+      end
+    end
+
+  private
+
+    attr_accessor :user, :reason, :keep_original
+
+    def initialize(user, reason: "undeclared participants in frozen cohort", keep_original: false)
+      @user = user
+      @reason = reason
+      @keep_original = keep_original
+    end
+
+    def check_user_can_be_archived!
+      if users_excluded_roles.any?
+        raise ArchiveError, "User #{user.id} has excluded roles: #{users_excluded_roles.join(',')}"
+      elsif user_profiles_are_not_all_archivable?
+        raise ArchiveError, "User #{user.id} has profiles in non-frozen cohorts"
```

Comment:
maybe this error messages is not meaningful when the participant has a profile in a frozen cohort that cannot be archived (unfinished 2021 for instance)

Reply:
Yes it's just an indication as previously I checked things separately but with the checks added to ParticipantProfile we only get true/false - but I thought at least it gives you an idea where to look